reasoner_util/__init__.py

```diff
@@ -35,3 +38,34 @@ def normalize_ids(curies: List[str]) -> List[str]:
     normalized_ids = get_preferred_ids(curies)
     # TODO perform function that strips descendents from the curies list.
     return normalized_ids
+
+
+def normalize_qcategories(catagories: List[str]) -> List[str]:
+    """Normalize a list of catagories by stripping all descendents"""
+    normalized_catagories = strip_descendants(catagories)
+    return normalized_catagories
+
+
+def normalize_predicates(predicates: List[str]) -> List[str]:
+    """Normalize a list of predicates by stripping all descendents"""
+    normalized_predicates = strip_descendants(predicates)
+    return normalized_predicates
+
+
+tk = Toolkit()
+
+
+def strip_descendants(item_list: List[str]) -> List[str]:
+    """strip descendants of biolink catagories or predicates"""
+    working_list = item_list.copy()
+    for x_item in item_list:
+        descendants = tk.get_descendants(
+            x_item,
+            reflexive=False,
+            formatted=True,
+        )
+        for desnd in descendants:
+            if desnd in working_list:
+                working_list.remove(desnd)
```

Comment:
This whole function has worst-case time complexity that's quadratic in the number of items N (`remove()` is linear in the list length), and linear in the average number of descendants D, totaling O(N^2 D). I would have grabbed all of the descendants first (O(ND)) and subsequently filtered the list using a comprehension (O(N)), for a net O(ND). However, your best-case performance is better, and average-case could be up for grabs depending on the typical distribution of items (are there commonly a lot of descendants, and where are they in the list?). This is mostly academic; N and D are upper-bounded in the 10s-100s, so I don't expect speed to be an issue. The other consideration is readability, and I'll let you be the judge there.

Reply:
I just want to make sure I understand--So instead I should get a list of the descendants of all of the items in the input list and then filter out all of the items in the input list that are present in both?

Reply:
That's my alternative algorithm in a nutshell, yes.